### DISCUSSION TITLE
Feature request: UI organisation-unit-tree missing code

### DISCUSSION THREAD

**Discussion by simonadomnisoru** [dhis2/notes#322] [Q&A]
Hello,

According to the organisation-unit-tree [UI docs](https://ui.dhis2.nu/demo/?path=/story/forms-organisation-unit-tree--collapsed), `onChange` is called with the following object ` { id: string, displayName: string, path: string, checked: boolean, selected: string[] }`.  Is it possible to add the organization unit's `code` to the object? The `code` is a piece of important information and it will be very useful to be able to retrieve it when the user changes the organization. Should I create a Jira issue for this?

Thank you!

**Comment by ghost** [comment]
@Mohammer5 knows more about this component. And I think similar requests have come up in the past for other components as well. I'm not sure what the long term vision is on these kinds of additions across the lib, @varl?

**Reply by Mohammer5** [reply]
I've been thinking about this for quite some time now and still think the best way to achieve this without having to add too many specific props to the component or fields & filters to the api request is by allowing the app to modify the request behavior (see dhis2/ui#681).

**Comment by khalilurrrahmanridoykhan** [comment]
Current status: the tree callback still does **not** include the organisation unit `code`. Its documented payload remains `{ id, displayName, path, checked, selected }`, as shown in the current [`OrganisationUnitTree` prop definition](https://github.com/dhis2/ui/blob/fc2740b54a672583df936d1f53b25a1620aafde9/components/organisation-unit-tree/src/organisation-unit-tree/organisation-unit-tree.js#L126-L128).

A reliable application-level solution is to use the callback's `id` as the identifier and obtain the code through the DHIS2 API/data engine, for example:

```js
const query = {
    orgUnit: {
        resource: 'organisationUnits',
        id: ({ id }) => id,
        params: { fields: ['id', 'code', 'displayName', 'path'] },
    },
}
```

Then execute that query with the `id` received by `onChange`. If many selections are expected, fetch the required organisation units once and keep an `id -> code` map rather than making a request on every click.

Adding `code` directly to the callback would change the component's public payload and require the tree's underlying queries to fetch that field. Therefore, a tracked feature request is still appropriate if the goal is native callback support; until then, joining by the stable organisation-unit ID avoids depending on the displayed label.
